# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## DataFrame Basics III

### Exercise 8: apply(), MultiIndex and Strings

In [1]:
import pandas as pd

In [2]:
cars = pd.read_csv("cars.csv")

In [3]:
cars

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
5    15.0          8         429.0       198.0    4341          10.0   
6    14.0          8         454.0       220.0    4354           9.0   
7    14.0          8         440.0       215.0    4312           8.5   
8    14.0          8         455.0       225.0    4425          10.0   
9    15.0          8         390.0       190.0    3850           8.5   
10   15.0          8         383.0       170.0    3563          10.0   
11   14.0          8         340.0       160.0    3609           8.0   
12   15.0          8         400.0       150.0    3761           9.5   
13   14.0          8         455.0       225.0    3086          10.0   
14   24.0          4         113.0        95.0    2372          15.0   
15   22.0          6         198.0        95.0    2833          15.5   
16   18.0          6         199.0        97.0    2774          15.5   
17   21.0          6         200.0        85.0    2587          16.0   
18   27.0          4          97.0        88.0    2130          14.5   
19   26.0          4          97.0        46.0    1835          20.5   
20   25.0          4         110.0        87.0    2672          17.5   
21   24.0          4         107.0        90.0    2430          14.5   
22   25.0          4         104.0        95.0    2375          17.5   
23   26.0          4         121.0       113.0    2234          12.5   
24   21.0          6         199.0        90.0    2648          15.0   
25   10.0          8         360.0       215.0    4615          14.0   
26   10.0          8         307.0       200.0    4376          15.0   
27   11.0          8         318.0       210.0    4382          13.5   
28    9.0          8         304.0       193.0    4732          18.5   
29   27.0          4          97.0        88.0    2130          14.5   
..    ...        ...           ...         ...     ...           ...   
368  27.0          4         112.0        88.0    2640          18.6   
369  34.0          4         112.0        88.0    2395          18.0   
370  31.0          4         112.0        85.0    2575          16.2   
371  29.0          4         135.0        84.0    2525          16.0   
372  27.0          4         151.0        90.0    2735          18.0   
373  24.0          4         140.0        92.0    2865          16.4   
374  23.0          4         151.0         NaN    3035          20.5   
375  36.0          4         105.0        74.0    1980          15.3   
376  37.0          4          91.0        68.0    2025          18.2   
377  31.0          4          91.0        68.0    1970          17.6   
378  38.0          4         105.0        63.0    2125          14.7   
379  36.0          4          98.0        70.0    2125          17.3   
380  36.0          4         120.0        88.0    2160          14.5   
381  36.0          4         107.0        75.0    2205          14.5   
382  34.0          4         108.0        70.0    2245          16.9   
383  38.0          4          91.0        67.0    1965          15.0   
384  32.0          4          91.0        67.0    1965          15.7   
385  38.0          4          91.0        67.0    1995          16.2   
386  25.0          6         181.0       110.0    2945          16.4   
387  38.0          6         262.0        85.0    3015          17.0   
388  26.0          4         156.0        92.0    2585          14.5   
389  22.0          6         232.0       112.0    2835          14.7   
390  32.0          4         144.0        96.0    2665          13.9   
391  36.0          4         135.0        84.0

61. __Calculate__ for all __numerical columns__ the __range between highest and lowest__ value! __Fill in the gaps!__ The range for __mpg__ is...?

In [4]:
cars.iloc[:, :-2].apply(lambda x: x.max() - x.min(), axis = 0)

mpg               37.6
cylinders          5.0
displacement     387.0
horsepower       184.0
weight          3527.0
acceleration      16.8
model_year        12.0
dtype: float64

The range for mpg is ... 37.6!

62. Create a __MultiIndex__ for the __cars__ DataFrame with the columns __"model_year"__ (outer level) and __"origin"__ (inner level). __Save__ the change!

In [4]:
cars.set_index(["model_year", "origin"], inplace = True)

63. __Sort__ both index levels in __ascending__ order! __Save__ the change!

In [5]:
cars.sort_index(ascending = True, inplace = True )

__Inspect__ Cars! What is the __name__ of the __second__ car?

In [6]:
cars

mpg  cylinders  displacement  horsepower  weight  \
model_year origin                                                      
70         europe  26.0          4          97.0        46.0    1835   
           europe  25.0          4         110.0        87.0    2672   
           europe  24.0          4         107.0        90.0    2430   
           europe  25.0          4         104.0        95.0    2375   
           europe  26.0          4         121.0       113.0    2234   
           japan   24.0          4         113.0        95.0    2372   
           japan   27.0          4          97.0        88.0    2130   
           usa     18.0          8         307.0       130.0    3504   
           usa     15.0          8         350.0       165.0    3693   
           usa     18.0          8         318.0       150.0    3436   
           usa     16.0          8         304.0       150.0    3433   
           usa     17.0          8         302.0       140.0    3449   
           usa     15.0          8         429.0       198.0    4341   
           usa     14.0          8         454.0       220.0    4354   
           usa     14.0          8         440.0       215.0    4312   
           usa     14.0          8         455.0       225.0    4425   
           usa     15.0          8         390.0       190.0    3850   
           usa     15.0          8         383.0       170.0    3563   
           usa     14.0          8         340.0       160.0    3609   
           usa     15.0          8         400.0       150.0    3761   
           usa     14.0          8         455.0       225.0    3086   
           usa     22.0          6         198.0        95.0    2833   
           usa     18.0          6         199.0        97.0    2774   
           usa     21.0          6         200.0        85.0    2587   
           usa     21.0          6         199.0        90.0    2648   
           usa     10.0          8         360.0       215.0    4615   
           usa     10.0          8         307.0       200.0    4376   
           usa     11.0          8         318.0       210.0    4382   
           usa      9.0          8         304.0       193.0    4732   
71         europe  28.0          4         116.0        90.0    2123   
...                 ...        ...           ...         ...     ...   
82         europe  44.0          4          97.0        52.0    2130   
           japan   37.0          4          91.0        68.0    2025   
           japan   31.0          4          91.0        68.0    1970   
           japan   36.0          4         120.0        88.0    2160   
           japan   36.0          4         107.0        75.0    2205   
           japan   34.0          4         108.0        70.0    2245   
           japan   38.0          4          91.0        67.0    1965   
           japan   32.0          4          91.0        67.0    1965   
           japan   38.0          4          91.0        67.0    1995   
           japan   32.0          4         144.0        96.0    2665   
           usa     28.0          4         112.0        88.0    2605   
           usa     27.0          4         112.0        88.0    2640   
           usa     34.0          4         112.0        88.0    2395   
           usa     31.0          4         112.0        85.0    2575   
           usa     29.0          4         135.0        84.0    2525   
           usa     27.0          4         151.0        90.0    2735   
           usa     24.0          4         140.0        92.0    2865   
           usa     23.0          4         151.0         NaN    3035   
           usa     38.0          4         105.0        63.0    2125   
           usa     36.0          4          98.0        70.0    2125   
           usa     25.0          6         181.0       110.0    2945   
           usa     38.0          6         262.0        85.0    3015   
           usa     26.0          4         156.0        92.0

The name of the second car is ... peugeot 504!

64. __Select__ all cars with the "model_year" __75__!

In [7]:
cars.loc[75]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
origin                                                                    
europe  25.0          4          90.0        71.0    2223          16.5   
europe  29.0          4          90.0        70.0    1937          14.0   
europe  23.0          4         115.0        95.0    2694          15.0   
europe  23.0          4         120.0        88.0    2957          17.0   
europe  22.0          4         121.0        98.0    2945          14.5   
europe  25.0          4         121.0       115.0    2671          13.5   
japan   29.0          4          97.0        75.0    2171          16.0   
japan   24.0          4         134.0        96.0    2702          13.5   
japan   24.0          4         119.0        97.0    2545          17.0   
japan   33.0          4          91.0        53.0    1795          17.5   
usa     19.0          6         225.0        95.0    3264          16.0   
usa     18.0          6         250.0       105.0    3459          16.0   
usa     15.0          6         250.0        72.0    3432          21.0   
usa     15.0          6         250.0        72.0    3158          19.5   
usa     16.0          8         400.0       170.0    4668          11.5   
usa     15.0          8         350.0       145.0    4440          14.0   
usa     16.0          8         318.0       150.0    4498          14.5   
usa     14.0          8         351.0       148.0    4657          13.5   
usa     17.0          6         231.0       110.0    3907          21.0   
usa     16.0          6         250.0       105.0    3897          18.5   
usa     15.0          6         258.0       110.0    3730          19.0   
usa     18.0          6         225.0        95.0    3785          19.0   
usa     21.0          6         231.0       110.0    3039          15.0   
usa     20.0          8         262.0       110.0    3221          13.5   
usa     13.0          8         302.0       129.0    3169          12.0   
usa     23.0          4         140.0        83.0    2639          17.0   
usa     20.0          6         232.0       100.0    2914          16.0   
usa     23.0          4         140.0        78.0    2592          18.5   
usa     18.0          6         171.0        97.0    2984          14.5   
usa     19.0          6         232.0        90.0    3211          17.0   

                             name  
origin                             
europe                  vw dasher  
europe                  vw rabbit  
europe                 audi 100ls  
europe                peugeot 504  
europe                volvo 244dl  
europe                  saab 99le  
japan              toyota corolla  
japan               toyota corona  
japan                  datsun 710  
japan            honda civic cvcc  
usa       plymouth valiant custom  
usa                chevrolet nova  
usa               mercury monarch  
usa                 ford maverick  
usa              pontiac catalina  
usa             chevrolet bel air  
usa           plymouth grand fury  
usa                      ford ltd  
usa                 buick century  
usa     chevrolet chevelle malibu  
usa                   amc matador  
usa                 plymouth fury  
usa                 buick skyhawk  
usa           chevrolet monza 2+2  
usa               ford mustang ii  
usa                    ford pinto  
usa                   amc gremlin  
usa                 pontiac astro  
usa                    ford pinto  
usa                     amc pacer

65. __Select__ all cars with the "model_year" __76__ from __japan__! __How many__ cars do we get?

In [8]:
cars.loc[(76, "japan")]

mpg  cylinders  displacement  horsepower  weight  \
model_year origin                                                      
76         japan   33.0          4          91.0        53.0    1795   
           japan   32.0          4          85.0        70.0    1990   
           japan   28.0          4          97.0        75.0    2155   
           japan   19.0          6         156.0       108.0    2930   

                   acceleration            name  
model_year origin                                
76         japan           17.4     honda civic  
           japan           17.0    datsun b-210  
           japan           16.4  toyota corolla  
           japan           15.5  toyota mark ii

There are ... 4 cars!

66. __Select__ all cars with the "model_year" __77__ and __79__  from __japan__! And only the columns __"mpg"__ and __"name"__! What is the __mpg__ of __datsun 210__?

In [9]:
cars.loc[([77,79], "japan"), ["mpg", "name"]]

mpg                     name
model_year origin                               
77         japan   31.5        honda accord cvcc
           japan   33.5    datsun f-10 hatchback
           japan   26.0  toyota corolla liftback
           japan   30.0                subaru dl
           japan   22.0               datsun 810
           japan   21.5               mazda rx-4
79         japan   34.1         mazda glc deluxe
           japan   31.8               datsun 210

The mpg of datsun 210 is ... 31.8!

67. __Swap the levels__ of the MultiIndex and then __sort__ the new MultiIndex in __ascending__ order! __Reassign cars__! __Fill in the gaps__!

In [10]:
cars = cars.swaplevel().sort_index(ascending = True)

In [11]:
cars

mpg  cylinders  displacement  horsepower  weight  \
origin model_year                                                      
europe 70          26.0          4          97.0        46.0    1835   
       70          25.0          4         110.0        87.0    2672   
       70          24.0          4         107.0        90.0    2430   
       70          25.0          4         104.0        95.0    2375   
       70          26.0          4         121.0       113.0    2234   
       71          28.0          4         116.0        90.0    2123   
       71          30.0          4          79.0        70.0    2074   
       71          30.0          4          88.0        76.0    2065   
       71          27.0          4          97.0        60.0    1834   
       72          23.0          4          97.0        54.0    2254   
       72          18.0          4         121.0       112.0    2933   
       72          22.0          4         121.0        76.0    2511   
       72          21.0          4         120.0        87.0    2979   
       72          26.0          4          96.0        69.0    2189   
       73          26.0          4          97.0        46.0    1950   
       73          26.0          4          98.0        90.0    2265   
       73          29.0          4          68.0        49.0    1867   
       73          24.0          4         116.0        75.0    2158   
       73          20.0          4         114.0        91.0    2582   
       73          19.0          4         121.0       112.0    2868   
       73          24.0          4         121.0       110.0    2660   
       74          29.0          4          98.0        83.0    2219   
       74          26.0          4          79.0        67.0    1963   
       74          26.0          4          97.0        78.0    2300   
       74          24.0          4          90.0        75.0    2108   
       74          26.0          4         116.0        75.0    2246   
       74          31.0          4          79.0        67.0    2000   
       75          25.0          4          90.0        71.0    2223   
       75          29.0          4          90.0        70.0    1937   
       75          23.0          4         115.0        95.0    2694   
...                 ...        ...           ...         ...     ...   
usa    81          23.5          6         173.0       110.0    2725   
       81          30.0          4         135.0        84.0    2385   
       81          39.0          4          86.0        64.0    1875   
       81          34.7          4         105.0        63.0    2215   
       81          34.4          4          98.0        65.0    2045   
       81          29.9          4          98.0        65.0    2380   
       81          22.4          6         231.0       110.0    3415   
       81          26.6          8         350.0       105.0    3725   
       81          20.2          6         200.0        88.0    3060   
       81          17.6          6         225.0        85.0    3465   
       82          28.0          4         112.0        88.0    2605   
       82          27.0          4         112.0        88.0    2640   
       82          34.0          4         112.0        88.0    2395   
       82          31.0          4         112.0        85.0    2575   
       82          29.0          4         135.0        84.0    2525   
       82          27.0          4         151.0        90.0    2735   
       82          24.0          4         140.0        92.0    2865   
       82          23.0          4         151.0         NaN    3035   
       82          38.0          4         105.0        63.0    2125   
       82          36.0          4          98.0        70.0    2125   
       82          25.0          6         181.0       110.0    2945   
       82          38.0          6         262.0        85.0    3015   
       82          26.0          4         156.0        92.0

68. __Reset__ the __index__ of __cars__ and create a RangeIndex! __Save__ the change!

In [12]:
cars.reset_index(inplace = True)

In [13]:
cars

origin  model_year   mpg  cylinders  displacement  horsepower  weight  \
0    europe          70  26.0          4          97.0        46.0    1835   
1    europe          70  25.0          4         110.0        87.0    2672   
2    europe          70  24.0          4         107.0        90.0    2430   
3    europe          70  25.0          4         104.0        95.0    2375   
4    europe          70  26.0          4         121.0       113.0    2234   
5    europe          71  28.0          4         116.0        90.0    2123   
6    europe          71  30.0          4          79.0        70.0    2074   
7    europe          71  30.0          4          88.0        76.0    2065   
8    europe          71  27.0          4          97.0        60.0    1834   
9    europe          72  23.0          4          97.0        54.0    2254   
10   europe          72  18.0          4         121.0       112.0    2933   
11   europe          72  22.0          4         121.0        76.0    2511   
12   europe          72  21.0          4         120.0        87.0    2979   
13   europe          72  26.0          4          96.0        69.0    2189   
14   europe          73  26.0          4          97.0        46.0    1950   
15   europe          73  26.0          4          98.0        90.0    2265   
16   europe          73  29.0          4          68.0        49.0    1867   
17   europe          73  24.0          4         116.0        75.0    2158   
18   europe          73  20.0          4         114.0        91.0    2582   
19   europe          73  19.0          4         121.0       112.0    2868   
20   europe          73  24.0          4         121.0       110.0    2660   
21   europe          74  29.0          4          98.0        83.0    2219   
22   europe          74  26.0          4          79.0        67.0    1963   
23   europe          74  26.0          4          97.0        78.0    2300   
24   europe          74  24.0          4          90.0        75.0    2108   
25   europe          74  26.0          4         116.0        75.0    2246   
26   europe          74  31.0          4          79.0        67.0    2000   
27   europe          75  25.0          4          90.0        71.0    2223   
28   europe          75  29.0          4          90.0        70.0    1937   
29   europe          75  23.0          4         115.0        95.0    2694   
..      ...         ...   ...        ...           ...         ...     ...   
368     usa          81  23.5          6         173.0       110.0    2725   
369     usa          81  30.0          4         135.0        84.0    2385   
370     usa          81  39.0          4          86.0        64.0    1875   
371     usa          81  34.7          4         105.0        63.0    2215   
372     usa          81  34.4          4          98.0        65.0    2045   
373     usa          81  29.9          4          98.0        65.0    2380   
374     usa          81  22.4          6         231.0       110.0    3415   
375     usa          81  26.6          8         350.0       105.0    3725   
376     usa          81  20.2          6         200.0        88.0    3060   
377     usa          81  17.6          6         225.0        85.0    3465   
378     usa          82  28.0          4         112.0        88.0    2605   
379     usa          82  27.0          4         112.0        88.0    2640   
380     usa          82  34.0          4         112.0        88.0    2395   
381     usa          82  31.0          4         112.0        85.0    2575   
382     usa          82  29.0          4         135.0        84.0    2525   
383     usa          82  27.0          4         151.0        90.0    2735   
384     usa          82  24.0          4         140.0        92.0    2865   
385     usa          82  23.0          4         151.0         NaN    3035   
386     usa          82  38.0          4         105.0        63.0    2125   
387     usa          82  36

69. __Convert__ all elements in the __origin__ column to __titlecase__! __Overwrite__ the __origin__ column!

In [14]:
cars.origin = cars.origin.str.title()

In [15]:
cars

origin  model_year   mpg  cylinders  displacement  horsepower  weight  \
0    Europe          70  26.0          4          97.0        46.0    1835   
1    Europe          70  25.0          4         110.0        87.0    2672   
2    Europe          70  24.0          4         107.0        90.0    2430   
3    Europe          70  25.0          4         104.0        95.0    2375   
4    Europe          70  26.0          4         121.0       113.0    2234   
5    Europe          71  28.0          4         116.0        90.0    2123   
6    Europe          71  30.0          4          79.0        70.0    2074   
7    Europe          71  30.0          4          88.0        76.0    2065   
8    Europe          71  27.0          4          97.0        60.0    1834   
9    Europe          72  23.0          4          97.0        54.0    2254   
10   Europe          72  18.0          4         121.0       112.0    2933   
11   Europe          72  22.0          4         121.0        76.0    2511   
12   Europe          72  21.0          4         120.0        87.0    2979   
13   Europe          72  26.0          4          96.0        69.0    2189   
14   Europe          73  26.0          4          97.0        46.0    1950   
15   Europe          73  26.0          4          98.0        90.0    2265   
16   Europe          73  29.0          4          68.0        49.0    1867   
17   Europe          73  24.0          4         116.0        75.0    2158   
18   Europe          73  20.0          4         114.0        91.0    2582   
19   Europe          73  19.0          4         121.0       112.0    2868   
20   Europe          73  24.0          4         121.0       110.0    2660   
21   Europe          74  29.0          4          98.0        83.0    2219   
22   Europe          74  26.0          4          79.0        67.0    1963   
23   Europe          74  26.0          4          97.0        78.0    2300   
24   Europe          74  24.0          4          90.0        75.0    2108   
25   Europe          74  26.0          4         116.0        75.0    2246   
26   Europe          74  31.0          4          79.0        67.0    2000   
27   Europe          75  25.0          4          90.0        71.0    2223   
28   Europe          75  29.0          4          90.0        70.0    1937   
29   Europe          75  23.0          4         115.0        95.0    2694   
..      ...         ...   ...        ...           ...         ...     ...   
368     Usa          81  23.5          6         173.0       110.0    2725   
369     Usa          81  30.0          4         135.0        84.0    2385   
370     Usa          81  39.0          4          86.0        64.0    1875   
371     Usa          81  34.7          4         105.0        63.0    2215   
372     Usa          81  34.4          4          98.0        65.0    2045   
373     Usa          81  29.9          4          98.0        65.0    2380   
374     Usa          81  22.4          6         231.0       110.0    3415   
375     Usa          81  26.6          8         350.0       105.0    3725   
376     Usa          81  20.2          6         200.0        88.0    3060   
377     Usa          81  17.6          6         225.0        85.0    3465   
378     Usa          82  28.0          4         112.0        88.0    2605   
379     Usa          82  27.0          4         112.0        88.0    2640   
380     Usa          82  34.0          4         112.0        88.0    2395   
381     Usa          82  31.0          4         112.0        85.0    2575   
382     Usa          82  29.0          4         135.0        84.0    2525   
383     Usa          82  27.0          4         151.0        90.0    2735   
384     Usa          82  24.0          4         140.0        92.0    2865   
385     Usa          82  23.0          4         151.0         NaN    3035   
386     Usa          82  38.0          4         105.0        63.0    2125   
387     Usa          82  36

Now, we want to __extract the manufacturer names__ from the column __"name"__. <br>
70. __Split__ the strings in the column __"name"__ on the very __first whitespace__ and __expand__ the splitted strings into __separate columns__! __Save__ new DataFrame in the variable __split__! __Fill in the gaps!__

In [16]:
split = cars.name.str.split(pat = " ", n = 1, expand = True)

__Inspect__!

In [17]:
split

0                       1
0            vw       1131 deluxe sedan
1       peugeot                     504
2          audi                  100 ls
3          saab                     99e
4           bmw                    2002
5          opel                    1900
6       peugeot                     304
7          fiat                    124b
8            vw               model 111
9            vw                  type 3
10        volvo               145e (sw)
11           vw                411 (sw)
12      peugeot                504 (sw)
13      renault                 12 (sw)
14           vw            super beetle
15         fiat         124 sport coupe
16         fiat                     128
17         opel                   manta
18         audi                   100ls
19        volvo                   144ea
20         saab                    99le
21         audi                     fox
22           vw                  dasher
23         opel                   manta
24         fiat                     128
25         fiat                  124 tc
26         fiat                    x1.9
27           vw                  dasher
28           vw                  rabbit
29         audi                   100ls
..          ...                     ...
368   chevrolet                citation
369    plymouth                 reliant
370    plymouth                   champ
371    plymouth               horizon 4
372        ford               escort 4w
373        ford               escort 2h
374       buick                 century
375  oldsmobile              cutlass ls
376        ford              granada gl
377    chrysler           lebaron salon
378   chevrolet                cavalier
379   chevrolet          cavalier wagon
380   chevrolet         cavalier 2-door
381     pontiac      j2000 se hatchback
382       dodge                aries se
383     pontiac                 phoenix
384        ford         fairmont futura
385         amc              concord dl
386    plymouth           horizon miser
387     mercury                  lynx l
388       buick         century limited
389  oldsmobile  cutlass ciera (diesel)
390    chrysler       lebaron medallion
391        ford               granada l
392       dodge             charger 2.2
393   chevrolet                  camaro
394        ford              mustang gl
395       dodge                 rampage
396        ford                  ranger
397   chevrolet                    s-10

[398 rows x 2 columns]

71. __Select__ the __first column__ and determine the __frequency of unique values__. __Fill in the gaps__! Which manufacturer is __most frequent__?

In [19]:
split[0].value_counts()

ford             51
chevrolet        47
plymouth         31
dodge            28
amc              28
toyota           26
datsun           23
vw               22
buick            17
pontiac          16
honda            13
mazda            12
mercury          12
oldsmobile       10
fiat              8
peugeot           8
audi              7
chrysler          6
volvo             6
renault           5
saab              4
subaru            4
opel              4
mercedes-benz     3
cadillac          2
bmw               2
ih                1
triumph           1
nissan            1
Name: 0, dtype: int64

There are 51 cars from the manufacturer ... Ford!

72. Let´s __doublecheck__ the result from above. Check __how many__ car names in the column "names" __contain "ford"__!

In [20]:
cars.name.str.contains("ford").sum()

51

It´s ... again 51! It seems that the very first word in the column "name" is always the manufacturer name.

# Well Done!

------------------------------

# Hints (Spoiler!)

61. apply() method, maximum - minimum

62. set_index() method, pass list with column labels

63. sort_index() method

64. pass outer index labels to loc operator

65. cars.loc[(outer index labels, inner index labels)]

66. cars.loc[(outer index labels, inner index labels), column labels]

67. methods swaplevel() and sort_index()

68. reset_index() method

69. string(str)-method title()

70. string(str)-method split(), n = 1, 

71. value_counts() method 

72. string(str)-method contains()